File: scripts/test-v2-mt5-terminal-clock.mjs
#!/usr/bin/env node

import assert from "node:assert/strict";
import { spawn, spawnSync } from "node:child_process";
import fs from "node:fs";
import { createServer } from "node:http";
import net from "node:net";
import path from "node:path";
import { pathToFileURL } from "node:url";
import { compileTypescriptModules } from "./v2-baseline/compile-typescript-modules.mjs";
import { evaluateV2Mt5TerminalContractAgreement } from "./v2-mt5-terminal-contract-agreement.mjs";

const workspace = process.cwd();
const authority = Object.freeze({
  executionAuthority: "none",
  brokerAuthority: "none",
  readinessOverrideAuthority: "none"
});
const python = spawnSync("python", [path.join("scripts", "test-v2-mt5-terminal-clock.py")], {
  cwd: workspace,
  encoding: "utf8"
});
assert.equal(python.status, 0, python.stderr || python.stdout);
const probeSource = fs.readFileSync(path.join(workspace, "mt5", "GoTraderClockProbe.mq5"), "utf8");
for (const forbiddenCall of [
  /\bCTrade\b/,
  /\bOrderSend\s*\(/,
  /\bAccountInfo\w*\s*\(/,
  /\bPosition\w*\s*\(/,
  /\bHistoryDeal\w*\s*\(/,
  /\bWebRequest\s*\(/,
  /\bSocket\w*\s*\(/,
  /^\s*#import\b/m
]) assert.doesNotMatch(probeSource, forbiddenCall);

const outRoot = path.join(workspace, ".gotrader", "v2-mt5-terminal-clock-test");
const sourceFiles = [
  "src/lib/v2/authority/v2Authority.ts",
  "src/lib/v2/serialization/canonicalSerialization.ts",
  "src/lib/v2/identity/v2IdentityTypes.ts",
  "src/lib/v2/identity/v2Identity.ts",
  "src/lib/v2/time/v2TimeNormalizationTypes.ts",
  "src/lib/v2/time/v2TimeNormalization.ts",
  "src/lib/v2/time/v2Mt5TerminalClockTypes.ts",
  "src/lib/v2/time/v2Mt5TerminalClock.ts",
  "src/lib/v2/time/v2Mt5UpstreamTimeContractTypes.ts",
  "src/lib/v2/time/v2Mt5UpstreamTimeContract.ts"
].map((file) => path.join(workspace, file));
compileTypescriptModules({ files: sourceFiles, outRoot });
const load = (name) => import(pathToFileURL(path.join(outRoot, `${name}.mjs`)).href);
const clock = await load("v2Mt5TerminalClock");
const contracts = await load("v2Mt5UpstreamTimeContract");
const identity = await load("v2Identity");

const observation = ({ offsetSeconds = 0, overrides = {} } = {}) => {
  const utcRaw = 1_800_000_000;
  const serverRaw = utcRaw + offsetSeconds;
  return Object.freeze({
    schemaId: "gotrader-mt5-terminal-clock-observation",
    version: "1.0.0",
    observationId: `ABCDEF12-100-${serverRaw}`,
    sequence: 100,
    probeInstanceId: "ABCDEF12",
    symbol: "USTECH",
    timeframe: "M5",
    captureDurationMs: 2,
    timeCurrentRaw: serverRaw,
    timeTradeServerRaw: serverRaw + 1,
    timeGmtRaw: utcRaw,
    timeLocalRaw: utcRaw - 14_400,
    timeGmtOffsetSeconds: 14_400,
    timeDaylightSavingsSeconds: -3_600,
    symbolTimeRaw: serverRaw,
    symbolTimeMscRaw: serverRaw * 1_000 + 250,
    latestBarOpenRaw: serverRaw - serverRaw % 300,
    latestBarIndex: 0,
    symbolSynchronized: true,
    tickReadSucceeded: true,
    barReadSucceeded: true,
    terminalBuild: 5836,
    authority,
    ...authority,
    ...overrides
  });
};
const compare = (terminalObservation, historicalDstPolicyVerified = false) => {
  const systemMs = terminalObservation.timeGmtRaw * 1_000;
  return clock.classifyV2Mt5TerminalClock({
    observation: terminalObservation,
    systemUtcBeforeMs: systemMs,
    systemUtcAfterMs: systemMs + 500,
    pythonTickRaw: terminalObservation.symbolTimeRaw,
    pythonTickMscRaw: terminalObservation.symbolTimeMscRaw,
    pythonLatestM5BarRaw: terminalObservation.latestBarOpenRaw,
    wrapperTickRaw: terminalObservation.symbolTimeRaw,
    wrapperLatestM5BarRaw: terminalObservation.latestBarOpenRaw,
    historicalDstPolicyVerified
  });
};

assert.equal(clock.validateV2Mt5TerminalClockObservation(observation()).status, "accepted");
assert.equal(clock.validateV2Mt5TerminalClockObservation({ ...observation(), orderData: [] }).status, "blocked");
assert.equal(clock.validateV2Mt5TerminalClockObservation({ ...observation(), brokerAuthority: "trade" }).status, "blocked");
assert.equal(clock.validateV2Mt5TerminalClockObservation({ ...observation(), sequence: "100" }).status, "blocked");
assert.equal(clock.validateV2Mt5TerminalClockObservation({ ...observation(), timeCurrentRaw: null }).status, "blocked");

const utc = compare(observation(), true);
assert.equal(utc.basisClassification, "verified_utc_epoch");
assert.equal(utc.phase2Eligible, true);
const wall = compare(observation({ offsetSeconds: 10_800 }));
assert.equal(wall.basisClassification, "verified_trade_server_wall_clock");
assert.equal(wall.pythonTransportBasis, "matches_symbol_quote_time");
assert.equal(wall.currentLiveTimeBasisVerified, true);
assert.equal(wall.historicalDstPolicyVerified, false);
assert.equal(wall.phase2Eligible, false);
assert.equal(wall.terminalObservedOffsetMinutes, 180);
assert.equal(wall.deltas.pythonCandleMinusLatestM5BarMs, 0);
const stale = clock.classifyV2Mt5TerminalClock({
  observation: observation({ offsetSeconds: 10_800 }),
  systemUtcBeforeMs: (1_800_000_000 + 500) * 1_000,
  systemUtcAfterMs: (1_800_000_000 + 501) * 1_000,
  pythonTickRaw: 1_800_010_800,
  pythonLatestM5BarRaw: 1_800_010_800
});
assert.equal(stale.currentLiveTimeBasisVerified, false);
assert.ok(stale.blockers.includes("terminal_observation_stale"));
const partial = clock.classifyV2Mt5TerminalClock({
  observation: observation({ offsetSeconds: 10_800, overrides: { barReadSucceeded: false } }),
  systemUtcBeforeMs: 1_800_000_000_000,
  systemUtcAfterMs: 1_800_000_000_500,
  pythonTickRaw: 1_800_010_800,
  pythonLatestM5BarRaw: 1_800_010_800
});
assert.equal(partial.phase2Eligible, false);
assert.ok(partial.blockers.includes("terminal_bar_unavailable"));
const staleQuote = compare(observation({
  offsetSeconds: 10_800,
  overrides: { timeTradeServerRaw: 1_800_010_800 + 1_800 }
}));
assert.equal(staleQuote.currentLiveTimeBasisVerified, false);
assert.ok(staleQuote.blockers.includes("terminal_quote_stale"));

const liveOnlyContract = Object.freeze({
  contractId: "gotrader-mt5-readonly-time-contract",
  version: "1.1.0",
  providerTimeBasis: "mt5_server_wall_clock",
  providerUtcOffsetMinutes: 180,
  dstPolicy: "fixed_offset",
  configurationSource: "operator_config",
  verificationStatus: "observed_candidate",
  verificationSources: ["terminal_clock_probe_current_live", "tick_candle_basis_comparison"],
  rawServerTime: 1_800_010_800,
  systemTimeUtc: "2027-01-15T08:00:00.000Z",
  rawLatestCandleTime: 1_800_010_700,
  tickCandleBasisAgreement: true,
  observationSummary: {
    observationCount: 0,
    acceptedObservationCount: 0,
    winterObservationCount: 0,
    summerObservationCount: 0,
    fixedOffsetObservationCount: 0
  },
  terminalProbeSchemaVersion: "1.0.0",
  terminalProbeObservationId: "ABCDEF12-100-1800010800",
  terminalProbeCapturedAt: "2027-01-15T08:00:00.000Z",
  terminalBasisClassification: "verified_trade_server_wall_clock",
  pythonTransportBasis: "matches_symbol_quote_time",
  terminalObservedOffsetMinutes: 180,
  terminalEvidenceStatus: "verified_current_live",
  terminalClockClassificationVersion: "1.0.0",
  timeVerificationScope: "current_live",
  currentLiveTimeBasisVerified: true,
  historicalDstPolicyVerified: false,
  phase2Eligible: false,
  strategySessionTimezone: "America/New_York",
  readOnly: true,
  marketDataOnly: true,
  blockers: [],
  warnings: [],
  authority,
  ...authority
});
const liveValidation = contracts.validateV2Mt5UpstreamTimeContract(liveOnlyContract);
assert.equal(liveValidation.status, "accepted");
assert.equal(liveValidation.phase2Eligible, false);
const acceptedAgreement = evaluateV2Mt5TerminalContractAgreement({
  responseStatus: 200,
  contract: { ...liveOnlyContract, sourceMethod: "upstream_http:/time-contract" },
  classification: wall
});
assert.equal(acceptedAgreement.status, "accepted");
assert.equal(acceptedAgreement.phase2Eligible, false);
const stubAgreement = evaluateV2Mt5TerminalContractAgreement({
  responseStatus: 200,
  contract: {
    ...liveOnlyContract,
    version: "0",
    verificationStatus: "unknown",
    currentLiveTimeBasisVerified: false,
    timeVerificationScope: "none",
    sourceMethod: "contract_stub:/time-contract"
  },
  classification: wall
});
assert.equal(stubAgreement.status, "blocked");
assert.ok(stubAgreement.blockers.includes("time_contract_version_mismatch"));
assert.ok(stubAgreement.blockers.includes("time_contract_stub_returned"));
assert.equal(contracts.validateV2Mt5UpstreamTimeContract({
  ...liveOnlyContract,
  verificationStatus: "verified"
}).status, "blocked");
assert.equal(contracts.validateV2Mt5UpstreamTimeContract({
  ...liveOnlyContract,
  phase2Eligible: true
}).status, "blocked");

const getFreePort = () => new Promise((resolve, reject) => {
  const server = net.createServer();
  server.once("error", reject);
  server.listen(0, "127.0.0.1", () => {
    const address = server.address();
    server.close(() => resolve(address.port));
  });
});
const mockUpstream = createServer((request, response) => {
  response.writeHead(request.url?.startsWith("/time-contract") ? 200 : 404, { "content-type": "application/json" });
  response.end(JSON.stringify(request.url?.startsWith("/time-contract") ? liveOnlyContract : { error: "not_found" }));
});
await new Promise((resolve) => mockUpstream.listen(0, "127.0.0.1", resolve));
const mockPort = mockUpstream.address().port;
const wrapperPort = await getFreePort();
const wrapper = spawn(process.execPath, [path.join(workspace, "scripts", "start-mt5-readonly-bridge.mjs")], {
  cwd: workspace,
  env: {
    ...process.env,
    MT5_READONLY_BRIDGE_HOST: "127.0.0.1",
    MT5_READONLY_BRIDGE_PORT: String(wrapperPort),
    MT5_READONLY_UPSTREAM_BASE_URL: `http://127.0.0.1:${mockPort}`,
    MT5_READONLY_UPSTREAM_TIMEOUT_MS: "1000"
  },
  stdio: ["ignore", "pipe", "pipe"]
});
try {
  let wrapped;
  for (let attempt = 0; attempt < 30 && !wrapped; attempt += 1) {
    await new Promise((resolve) => setTimeout(resolve, 100));
    wrapped = await fetch(`http://127.0.0.1:${wrapperPort}/time-contract`)
      .then((response) => response.ok ? response.json() : undefined)
      .catch(() => undefined);
  }
  assert.ok(wrapped, "MT5 wrapper did not expose the terminal-clock contract fixture.");
  assert.equal(wrapped.terminalProbeObservationId, liveOnlyContract.terminalProbeObservationId);
  assert.equal(wrapped.terminalBasisClassification, liveOnlyContract.terminalBasisClassification);
  assert.equal(wrapped.currentLiveTimeBasisVerified, true);
  assert.equal(wrapped.historicalDstPolicyVerified, false);
  assert.equal(wrapped.phase2Eligible, false);
  assert.equal(wrapped.timeVerificationScope, "current_live");
  assert.equal(wrapped.executionAuthority, "none");
} finally {
  wrapper.kill();
  await new Promise((resolve) => mockUpstream.close(resolve));
}

const source = identity.createV2SourceIdentity({
  sourceId: "mt5:USTECH:5m",
  provider: "mt5_read_only",
  requestedSymbol: "MNQ",
  brokerSymbol: "USTECH",
  sourceFingerprint: "legacy-fingerprint-unchanged",
  sourceKind: "mt5_read_only"
});
const identityInput = {
  source,
  timeframeFingerprints: { "5m": source.sourceFingerprint },
  dataWindowStart: "2027-01-15T07:45:00.000Z",
  dataWindowEnd: "2027-01-15T08:00:00.000Z",
  lastClosedCandle: "2027-01-15T08:00:00.000Z",
  candleCountByTimeframe: { "5m": 3 },
  timeNormalizationPolicyId: liveOnlyContract.contractId,
  timeNormalizationPolicyVersion: liveOnlyContract.version,
  timeContractId: liveOnlyContract.contractId,
  timeContractVersion: liveOnlyContract.version,
  timeContractVerificationStatus: liveOnlyContract.verificationStatus,
  terminalClockClassificationVersion: "1.0.0"
};
const currentIdentity = await identity.buildV2MarketDataIdentity({ ...identityInput, timeVerificationScope: "current_live" });
const historicalIdentity = await identity.buildV2MarketDataIdentity({ ...identityInput, timeVerificationScope: "historical" });
const volatileObservationIdentity = await identity.buildV2MarketDataIdentity({
  ...identityInput,
  timeVerificationScope: "current_live",
  terminalProbeObservationId: "DIFFERENT-OBSERVATION",
  terminalProbeCapturedAt: "2030-01-01T00:00:00.000Z",
  observationAgeMs: 99_999
});
assert.notEqual(currentIdentity.identityHash, historicalIdentity.identityHash);
assert.equal(currentIdentity.identityHash, volatileObservationIdentity.identityHash);

const output = {
  status: "passed",
  pythonReaderTests: "passed",
  utcClassification: utc.basisClassification,
  wallClockClassification: wall.basisClassification,
  currentLiveVerified: wall.currentLiveTimeBasisVerified,
  historicalDstVerified: wall.historicalDstPolicyVerified,
  phase2Eligible: wall.phase2Eligible,
  exactBarParityMs: wall.deltas.pythonCandleMinusLatestM5BarMs,
  staleRejected: stale.blockers.includes("terminal_observation_stale"),
  partialRejected: partial.blockers.includes("terminal_bar_unavailable"),
  staleQuoteRejected: staleQuote.blockers.includes("terminal_quote_stale"),
  identitySensitiveToVerificationScope: currentIdentity.identityHash !== historicalIdentity.identityHash,
  wrapperPassThrough: "passed",
  endpointAgreement: acceptedAgreement.status,
  stubMismatchBlocked: stubAgreement.status === "blocked",
  identityExcludesVolatileObservationFields: currentIdentity.identityHash === volatileObservationIdentity.identityHash,
  rawCandleArraysSerialized: false,
  sensitiveFieldsAbsent: true,
  mql5MutationCapabilitiesAbsent: true,
  partialCurrentBarCreatesEvidence: false,
  strategySessionTimezone: "America/New_York",
  authority
};
const serialized = JSON.stringify(output);
assert.doesNotMatch(serialized, /"candles"\s*:/i);
assert.doesNotMatch(serialized, /accountData|orderData|positionData|password|secret|apiKey|token/i);
console.log(JSON.stringify(output, null, 2));

File: scripts/test-v2-mt5-terminal-clock.py
#!/usr/bin/env python3
"""Deterministic tests for the read-only terminal clock observation reader."""

from __future__ import annotations

import json
import runpy
import tempfile
from datetime import datetime, timezone
from pathlib import Path

from v2_mt5_terminal_clock import (
    AUTHORITY,
    MAX_OBSERVATION_BYTES,
    TerminalClockObservationError,
    TerminalClockObservationRegistry,
    classify_terminal_clock,
    fnv1a32,
    observation_path,
    read_observation,
    validate_observation,
)


def expect_error(action, expected: str) -> None:
    try:
        action()
    except TerminalClockObservationError as error:
        assert expected in str(error), (expected, str(error))
        return
    raise AssertionError(f"Expected {expected}")


def fixture(*, utc_raw: int = 1_800_000_000, offset_seconds: int = 0) -> dict:
    server_raw = utc_raw + offset_seconds
    return {
        "schemaId": "gotrader-mt5-terminal-clock-observation",
        "version": "1.0.0",
        "observationId": f"ABCDEF12-100-{server_raw}",
        "sequence": 100,
        "probeInstanceId": "ABCDEF12",
        "symbol": "USTECH",
        "timeframe": "M5",
        "captureDurationMs": 2,
        "timeCurrentRaw": server_raw,
        "timeTradeServerRaw": server_raw + 1,
        "timeGmtRaw": utc_raw,
        "timeLocalRaw": utc_raw - 14_400,
        "timeGmtOffsetSeconds": 14_400,
        "timeDaylightSavingsSeconds": -3_600,
        "symbolTimeRaw": server_raw,
        "symbolTimeMscRaw": server_raw * 1000 + 250,
        "latestBarOpenRaw": server_raw - server_raw % 300,
        "latestBarIndex": 0,
        "symbolSynchronized": True,
        "tickReadSucceeded": True,
        "barReadSucceeded": True,
        "terminalBuild": 5836,
        "authority": AUTHORITY,
        **AUTHORITY,
    }


def compare(observation: dict, *, python_tick: int, python_bar: int, historical: bool = False) -> dict:
    system_ms = int(observation["timeGmtRaw"] * 1000)
    return classify_terminal_clock(
        observation=observation,
        system_utc_before_ms=system_ms,
        system_utc_after_ms=system_ms + 500,
        python_tick_raw=python_tick,
        python_tick_msc_raw=python_tick * 1000 + 250,
        python_latest_m5_bar_raw=python_bar,
        historical_dst_policy_verified=historical,
    )


valid = fixture()
validate_observation(valid)
expect_error(lambda: validate_observation({**valid, "accountData": {"login": 1}}), "sensitive_field_forbidden")
expect_error(lambda: validate_observation({**valid, "executionAuthority": "trade"}), "top_level_authority_invalid")

with tempfile.TemporaryDirectory() as directory:
    common = Path(directory) / "Common"
    data = str(Path(directory) / "Terminal" / "INSTANCE")
    path = observation_path(str(common), data)
    expect_error(
        lambda: read_observation(common_data_path=str(common), terminal_data_path=data),
        "file_missing",
    )
    path.parent.mkdir(parents=True)
    persisted = {**valid, "probeInstanceId": fnv1a32(data)}
    path.write_text(json.dumps(persisted), encoding="utf-8")
    registry = TerminalClockObservationRegistry.empty()
    loaded = read_observation(
        common_data_path=str(common),
        terminal_data_path=data,
        now_epoch_seconds=valid["timeGmtRaw"] + 1,
        registry=registry,
    )
    assert loaded["observationId"] == valid["observationId"]
    expect_error(
        lambda: read_observation(
            common_data_path=str(common),
            terminal_data_path=data,
            now_epoch_seconds=valid["timeGmtRaw"] + 1,
            registry=registry,
        ),
        "duplicate",
    )
    expect_error(
        lambda: read_observation(
            common_data_path=str(common),
            terminal_data_path=data,
            now_epoch_seconds=valid["timeGmtRaw"] + 500,
        ),
        "stale",
    )
    path.write_bytes(b"x" * (MAX_OBSERVATION_BYTES + 1))
    expect_error(
        lambda: read_observation(common_data_path=str(common), terminal_data_path=data),
        "file_size_invalid",
    )
    path.write_text("{", encoding="utf-8")
    expect_error(
        lambda: read_observation(common_data_path=str(common), terminal_data_path=data),
        "malformed",
    )
    path.write_text(json.dumps({**persisted, "schemaId": "invalid"}), encoding="utf-8")
    expect_error(
        lambda: read_observation(
            common_data_path=str(common),
            terminal_data_path=data,
            now_epoch_seconds=valid["timeGmtRaw"] + 1,
        ),
        "schema_invalid",
    )
    path.write_text(json.dumps({**persisted, "probeInstanceId": "00000000"}), encoding="utf-8")
    expect_error(
        lambda: read_observation(
            common_data_path=str(common),
            terminal_data_path=data,
            now_epoch_seconds=valid["timeGmtRaw"] + 1,
        ),
        "instance_mismatch",
    )

utc_observation = fixture(offset_seconds=0)
utc_result = compare(
    utc_observation,
    python_tick=utc_observation["symbolTimeRaw"],
    python_bar=utc_observation["latestBarOpenRaw"],
    historical=True,
)
assert utc_result["basisClassification"] == "verified_utc_epoch"
assert utc_result["phase2Eligible"] is True

wall_observation = fixture(offset_seconds=10_800)
wall_result = compare(
    wall_observation,
    python_tick=wall_observation["symbolTimeRaw"],
    python_bar=wall_observation["latestBarOpenRaw"],
)
assert wall_result["basisClassification"] == "verified_trade_server_wall_clock"
assert wall_result["currentLiveTimeBasisVerified"] is True
assert wall_result["historicalDstPolicyVerified"] is False
assert wall_result["phase2Eligible"] is False
assert wall_result["terminalObservedOffsetMinutes"] == 180

conflict_result = compare(
    wall_observation,
    python_tick=wall_observation["symbolTimeRaw"] + 600,
    python_bar=wall_observation["latestBarOpenRaw"] + 600,
)
assert conflict_result["basisClassification"] in {"insufficient_evidence", "conflicting_terminal_evidence"}
assert conflict_result["phase2Eligible"] is False

stale_quote_observation = {
    **wall_observation,
    "timeTradeServerRaw": wall_observation["timeCurrentRaw"] + 1_800,
}
stale_quote_result = compare(
    stale_quote_observation,
    python_tick=stale_quote_observation["symbolTimeRaw"],
    python_bar=stale_quote_observation["latestBarOpenRaw"],
)
assert stale_quote_result["currentLiveTimeBasisVerified"] is False
assert "terminal_quote_stale" in stale_quote_result["blockers"]

workspace = Path(__file__).resolve().parents[1]
upstream = runpy.run_path(str(workspace / "scripts" / "mt5-readonly-upstream.py"))
build_time_contract = upstream["build_time_contract"]
terminal_evidence = {
    "terminalProbeSchemaVersion": wall_observation["version"],
    "terminalProbeObservationId": wall_observation["observationId"],
    "terminalProbeCapturedAt": "2027-01-15T08:00:00Z",
    "terminalBasisClassification": wall_result["basisClassification"],
    "pythonTransportBasis": wall_result["pythonTransportBasis"],
    "terminalObservedOffsetMinutes": wall_result["terminalObservedOffsetMinutes"],
    "terminalEvidenceStatus": wall_result["terminalEvidenceStatus"],
    "timeVerificationScope": wall_result["verificationScope"],
    "currentLiveTimeBasisVerified": wall_result["currentLiveTimeBasisVerified"],
    "historicalDstPolicyVerified": wall_result["historicalDstPolicyVerified"],
    "terminalClockClassificationVersion": wall_result["classificationVersion"],
    "terminalProbeBlockers": wall_result["blockers"],
    "terminalProbeWarnings": wall_result["warnings"],
}
live_contract = build_time_contract(
    environment={},
    raw_tick_time=wall_observation["symbolTimeRaw"],
    raw_tick_time_msc=wall_observation["symbolTimeMscRaw"],
    raw_candle_time=wall_observation["latestBarOpenRaw"],
    system_time_utc=datetime.fromtimestamp(wall_observation["timeGmtRaw"], timezone.utc),
    terminal_build=5836,
    terminal_version="500.5836.28 Apr 2026",
    mt5_package_version="5.0.5735",
    terminal_evidence=terminal_evidence,
)
assert live_contract["verificationStatus"] == "observed_candidate"
assert live_contract["providerTimeBasis"] == "mt5_server_wall_clock"
assert live_contract["configurationSource"] == "provider_metadata"
assert live_contract["currentLiveTimeBasisVerified"] is True
assert live_contract["historicalDstPolicyVerified"] is False
assert live_contract["timeVerificationScope"] == "current_live"
assert live_contract["phase2Eligible"] is False

conflicting_contract = build_time_contract(
    environment={},
    raw_tick_time=wall_observation["symbolTimeRaw"],
    raw_tick_time_msc=wall_observation["symbolTimeMscRaw"],
    raw_candle_time=wall_observation["latestBarOpenRaw"],
    system_time_utc=datetime.fromtimestamp(wall_observation["timeGmtRaw"], timezone.utc),
    terminal_build=5836,
    terminal_version="500.5836.28 Apr 2026",
    mt5_package_version="5.0.5735",
    terminal_evidence={
        **terminal_evidence,
        "terminalBasisClassification": "conflicting_terminal_evidence",
        "terminalEvidenceStatus": "conflicting",
        "currentLiveTimeBasisVerified": False,
    },
)
assert conflicting_contract["verificationStatus"] == "unknown"
assert conflicting_contract["currentLiveTimeBasisVerified"] is False
assert conflicting_contract["historicalDstPolicyVerified"] is False
assert conflicting_contract["phase2Eligible"] is False

print(json.dumps({
    "status": "passed",
    "utcClassification": utc_result["basisClassification"],
    "wallClockClassification": wall_result["basisClassification"],
    "currentLiveVerified": wall_result["currentLiveTimeBasisVerified"],
    "historicalDstVerified": wall_result["historicalDstPolicyVerified"],
    "duplicateRejected": True,
    "staleRejected": True,
    "staleQuoteRejected": True,
    "oversizedRejected": True,
    "malformedRejected": True,
    "invalidSchemaRejected": True,
    "missingObservationRejected": True,
    "terminalInstanceMismatchRejected": True,
    "conflictingContractBlocked": conflicting_contract["verificationStatus"] == "unknown",
    "sensitiveFieldsRejected": True,
    "upstreamContractStatus": live_contract["verificationStatus"],
    "authority": AUTHORITY,
}, indent=2))
